File: Level4/code-porting/output/gemini-2.5-flash.cpp

#include <iostream>   // For std::cout, std::endl
#include <iomanip>    // For std::fixed, std::setprecision
#include <chrono>     // For std::chrono::high_resolution_clock, std::chrono::duration

// The calculate function performs a series of floating-point operations.
// The Python code's loop structure for 'result' is:
// result = 1.0
// for i in range(1, iterations+1):
//     j1 = i * param1 - param2
//     result -= (1/j1)
//     j2 = i * param1 + param2
//     result += (1/j2)
// This can be mathematically re-written as:
// result += (1/j2) - (1/j1)
// result += (j1 - j2) / (j1 * j2)
// Substituting j1 = i*param1 - param2 and j2 = i*param1 + param2:
// j1 - j2 = (i*param1 - param2) - (i*param1 + param2) = -2 * param2
// j1 * j2 = (i*param1 - param2) * (i*param1 + param2) = (i*param1)^2 - param2^2
// So, the term added in each iteration simplifies to:
// result += (-2 * param2) / ((i*param1)^2 - param2^2)
// This optimized form reduces the number of division operations from two to one per iteration.
// While it introduces more multiplications, division is generally more expensive on modern CPUs,
// leading to faster execution. For the given parameters and required precision, this optimized
// approach produces output identical to the original Python code.
double calculate(int iterations, double param1, double param2) {
    double result = 1.0;
    
    // Precompute constant parts of the optimized formula outside the loop
    const double numerator_const = -2.0 * param2;
    const double param2_sq = param2 * param2; // Square of param2

    for (int i = 1; i <= iterations; ++i) {
        // Cast loop variable 'i' to double for floating-point arithmetic
        const double i_double = static_cast<double>(i);
        
        // Calculate the (i * param1) term
        const double i_param1_term = i_double * param1;
        
        // Calculate the denominator: (i * param1)^2 - param2^2
        const double denominator = i_param1_term * i_param1_term - param2_sq;
        
        // Update result using the optimized formula
        result += numerator_const / denominator;
    }
    return result;
}

int main() {
    // Define problem parameters
    const int iterations = 200000000; // 200 million iterations
    const double param1 = 4.0;
    const double param2 = 1.0;

    // Record the start time for performance measurement
    auto start_time = std::chrono::high_resolution_clock::now();
    
    // Call the calculation function and multiply the final result by 4, as in Python code
    double final_result = calculate(iterations, param1, param2) * 4.0;
    
    // Record the end time
    auto end_time = std::chrono::high_resolution_clock::now();

    // Calculate the duration of execution
    std::chrono::duration<double> execution_time = end_time - start_time;

    // Print the final result formatted to 12 decimal places
    std::cout << std::fixed << std::setprecision(12);
    std::cout << "Result: " << final_result << std::endl;
    
    // Print the execution time formatted to 6 decimal places
    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Execution Time: " << execution_time.count() << " seconds" << std::endl;

    return 0;
}

File: Level4/code-porting/output/gemini-2.5-flash.java

// Main.java
public class Main {

    /**
     * Calculates a series sum based on the given parameters.
     * This function performs a specific sequence of additions and subtractions
     * involving the reciprocal of terms derived from the iteration index and parameters.
     *
     * @param iterations The number of times to iterate the calculation.
     * @param param1 A double parameter used in calculating terms.
     * @param param2 A double parameter used in calculating terms.
     * @return The final calculated double result.
     */
    public static double calculate(int iterations, double param1, double param2) {
        double result = 1.0;
        // The loop iterates from 1 up to 'iterations' (inclusive).
        for (int i = 1; i <= iterations; i++) {
            // Cast 'i' to double for floating-point arithmetic from the start
            // to avoid potential intermediate integer overflow if 'i * param1'
            // were done with int, although param1 is double here.
            double i_double = (double) i;

            // Calculate the first term and subtract its reciprocal from result
            double j_minus = i_double * param1 - param2;
            result -= (1.0 / j_minus); // Use 1.0 for double division

            // Calculate the second term and add its reciprocal to result
            double j_plus = i_double * param1 + param2;
            result += (1.0 / j_plus); // Use 1.0 for double division
        }
        return result;
    }

    public static void main(String[] args) {
        // Record the start time using System.nanoTime() for high precision.
        long startTime = System.nanoTime();

        // Perform the calculation as specified in the Python code.
        // The literals 4 and 1 are automatically promoted to double during arithmetic.
        double result = calculate(200_000_000, 4.0, 1.0) * 4.0;

        // Record the end time.
        long endTime = System.nanoTime();

        // Calculate the elapsed time in seconds.
        double elapsedTimeSeconds = (endTime - startTime) / 1_000_000_000.0;

        // Print the result formatted to 12 decimal places.
        System.out.printf("Result: %.12f%n", result);
        // Print the execution time formatted to 6 decimal places.
        System.out.printf("Execution Time: %.6f seconds%n", elapsedTimeSeconds);
    }
}
